server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,744 @@
+            final FieldInfos fieldInfos = reader.getFieldInfos();
+            for (Map.Entry<Integer, Long> field : visitor.fields.entrySet()) {
+                final String fieldName = fieldInfos.fieldInfo(field.getKey()).name;
+                final long fieldSize = (long) Math.ceil(field.getValue() * ratio);
+                stats.addStoredField(fieldName, fieldSize);
+            }
+        }
+    }
+
+    private static class TrackingSizeStoredFieldVisitor extends StoredFieldVisitor {
+        private final Map<Integer, Long> fields = new HashMap<>();
+
+        private void trackField(FieldInfo fieldInfo, int fieldLength) {
+            final int totalBytes = fieldLength + Long.BYTES; // a Long for bitsAndInfo
+            fields.compute(fieldInfo.number, (k, v) -> v == null ? totalBytes : v + totalBytes);
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES + value.length);
+        }
+
+        @Override
+        public void stringField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES + value.length);
+        }
+
+        @Override
+        public void intField(FieldInfo fieldInfo, int value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES);
+        }
+
+        @Override
+        public void longField(FieldInfo fieldInfo, long value) throws IOException {
+            trackField(fieldInfo, Long.BYTES);
+        }
+
+        @Override
+        public void floatField(FieldInfo fieldInfo, float value) throws IOException {
+            trackField(fieldInfo, Float.BYTES);
+        }
+
+        @Override
+        public void doubleField(FieldInfo fieldInfo, double value) throws IOException {
+            trackField(fieldInfo, Double.BYTES);
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            return Status.YES;
+        }
+    }
+
+    // The type bound should be DocValuesIterator instead of DocIdSetIterator.
+    private <DV extends DocIdSetIterator> DV iterateDocValues(int maxDocs,
+                                                              CheckedSupplier<DV, IOException> dvReader,
+                                                              CheckedConsumer<DV, IOException> valueAccessor) throws IOException {
+        // As we track the min/max positions of read bytes, we just visit the first and last values of the docValues iterator.
+        DV dv = dvReader.get();
+        int docID;
+        // Try to access the first value
+        if ((docID = dv.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+            valueAccessor.accept(dv);
+            // Try to access the last value by advancing the last doc.
+            // If the last doc doesn't have value then fallback to access the entire iterator
```

Comment:
can you add a TODO here to do a "binary search"-like approach as Adrien suggested. It might be reasonable for the last doc not to have a value (e.g. because the source did not have such a field, or because it's a delete tombstone doc), so it would be good to not only optimize this one case. Perhaps there's even a way to quickly lookup the document with highest doc id that has a value for this field (using exists query)?

Reply:
I implemented a binary-search like approach in https://github.com/elastic/elasticsearch/pull/74051/commits/cac127aeda26acbef8612f23a4c973b4f102e14a.